### PR TITLE
NetworkManager Support

### PR DESCRIPTION
**Pull request by AlvinSchiller**
PR for https://github.com/MiczFlor/RPi-Jukebox-RFID/issues/2184

- [x] Update setup
- [x] Update autohotspot based on dhcpcd
- [x] Add autohotspot based on NetworkManager
- [x] Fix autohotspot problems solved in v2 (https://github.com/MiczFlor/RPi-Jukebox-RFID/pull/2086)
- [x] Fix "Static IP" for NetworkManager
- [x] Fix "IPv6 disablement"
- [x] Update docs
- [x] fix path to boot files on bookworm (/boot/firmware)

Optional (maybe with a seperat PR)
- [ ] Add card actions for Wifi / Autohotspot handling (addresses https://github.com/MiczFlor/RPi-Jukebox-RFID/issues/1993)
- [ ] Add rpc call to force a Hotspot (add to gui and card actions)?

### CODE REVIEW

File: installation/routines/setup_autohotspot.sh [hunk 1]
```diff
@@ -1,119 +1,91 @@
-#!/usr/bin/env bash
```
Comment (s-martin) [inline installation/routines/setup_autohotspot.sh:1]: Don't we need this?

File: installation/routines/setup_autohotspot.sh [hunk 2]
```diff
@@ -1,119 +1,91 @@
...
+AUTOHOTSPOT_INTERFACES_CONF_FILE="/etc/network/interfaces"
+AUTOHOTSPOT_TARGET_PATH="/usr/bin/autohotspot"
+AUTOHOTSPOT_SERVICE="autohotspot.service"
+AUTOHOTSPOT_SERVICE_PATH="${SYSTEMD_PATH}/${AUTOHOTSPOT_SERVICE}"
+
+_is_service_enabled() {
```
Comment (s-martin) [inline installation/routines/setup_autohotspot.sh:6]: Put this function in the helpers file, because could be used by other features?

File: resources/autohotspot/dhcpcd/autohotspot.service [hunk 3]
```diff
@@ -0,0 +1,11 @@
+[Unit]
+Description=Automatically generates an wifi hotspot when a valid ssid is not in range
+After=multi-user.target
```
Comment (s-martin) [inline resources/autohotspot/dhcpcd/autohotspot.service:3]: Do we need network services available?

File: installation/routines/setup_autohotspot.sh [hunk 4]
```diff
@@ -1,119 +1,91 @@
...
-    sudo sed -i "s|%%USER%%|${CURRENT_USER}|g" "${cron_autohotspot_file}"
+_config_file_backup() {
+    # create flag file or copy present conf to orig file
+    # to correctly handling future deactivation of autohotspot
+    local config_file="$1"
+    local config_flag_file="${config_file}.remove"
+    local config_orig_file="${config_file}.orig"
+    if [ ! -f "${config_file}" ]; then
+        sudo touch "${config_flag_file}"
+    elif [ ! -f "${config_orig_file}" ] && [ ! -f "${config_flag_file}" ]; then
```
Comment (s-martin) [inline installation/routines/setup_autohotspot.sh:51]: If the orig file already exists and the autohotspot setup is run again the current config is not backup'd again.

File: installation/routines/setup_autohotspot_NetworkManager.sh [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    sudo sed -i "s|%%AUTOHOTSPOT_PASSWORD%%|${AUTOHOTSPOT_PASSWORD}|g" "${AUTOHOTSPOT_TARGET_PATH}"
+    sudo sed -i "s|%%AUTOHOTSPOT_IP%%|${AUTOHOTSPOT_IP}|g" "${AUTOHOTSPOT_TARGET_PATH}"
+    sudo sed -i "s|%%IP_WITHOUT_LAST_SEGMENT%%|${ip_without_last_segment}|g" "${AUTOHOTSPOT_TARGET_PATH}"
+    sudo sed -i "s|%%AUTOHOTSPOT_TIMER_NAME%%|${AUTOHOTSPOT_TIMER}|g" "${AUTOHOTSPOT_TARGET_PATH}"
+    
+    sudo chmod +x "${AUTOHOTSPOT_TARGET_PATH}"
+
+    sudo cp "${AUTOHOTSPOT_NETWORKMANAGER_RESOURCES_PATH}"/autohotspot.service "${AUTOHOTSPOT_SERVICE_PATH}"
+    sudo sed -i "s|%%AUTOHOTSPOT_SCRIPT%%|${AUTOHOTSPOT_TARGET_PATH}|g" "${AUTOHOTSPOT_SERVICE_PATH}"
+    #sudo chown root:root "${AUTOHOTSPOT_SERVICE_PATH}"
```
Comment (s-martin) [inline installation/routines/setup_autohotspot_NetworkManager.sh:31]: Is it intentional that it is commented out?

File: installation/routines/setup_autohotspot.sh (hunk 1 above)
Comment (AlvinSchiller) [reply installation/routines/setup_autohotspot.sh:1]: True, we could think about removing them on all scripts in the "routines" folder.
But that should be done with the reordering @pabera had started.

I would just keep this to have less irrelevant changes

File: installation/routines/setup_autohotspot_NetworkManager.sh (hunk 5 above)
Comment (AlvinSchiller) [reply installation/routines/setup_autohotspot_NetworkManager.sh:31]: yep, still a draft and some testing to do ;)

File: installation/routines/setup_autohotspot.sh (hunk 4 above)
Comment (AlvinSchiller) [reply installation/routines/setup_autohotspot.sh:51]: Yes, this is intentional (and also the current v2 behavior).
Just the very first file is backed up as it should be functional before we touch it and be restored after deinstallation.
If we change this file again on a subsequent run the changes shouldn't be interesting.

File: resources/autohotspot/dhcpcd/autohotspot.service (hunk 3 above)
Comment (AlvinSchiller) [reply resources/autohotspot/dhcpcd/autohotspot.service:3]: Was the state before and honstly just copied from the tutorial / script used as base.

File: installation/routines/setup_autohotspot.sh (hunk 1 above)
Comment (AlvinSchiller) [reply installation/routines/setup_autohotspot.sh:1]: Oh i got your questions wrong 🤦 
I thought you suggest to remove it. Id idn't see it was removed.

I will add it again, but the above remains valid ^^

File: installation/routines/setup_autohotspot.sh [hunk 6]
```diff
@@ -1,119 +1,91 @@
...
-}
 
 setup_autohotspot() {
     if [ "$ENABLE_AUTOHOTSPOT" == true ] ; then
-        run_with_log_frame _run_setup_autohotspot "Install AutoHotspot"
+        local installed=false
+        if [[ $(_is_dhcpcd_enabled) == true || "${CI_RUNNING}" == "true" ]]; then
+            run_with_log_frame _run_setup_autohotspot_dhcpcd "Install AutoHotspot dhcpcd"
+            installed=true
+        fi
```
Comment (pabera) [inline installation/routines/setup_autohotspot.sh:80]: Do we have this only for backwards compatibility? I don't really understand why we want to support dhcpcd and NetworkManager.

To reduce complexity, why don't we migrate to NetworkManager directly and ignore dhcpcd directly. Or do we need this for Bullseye, because it does not support NetworkManager?

File: installation/routines/setup_autohotspot.sh [hunk 7]
```diff
@@ -1,119 +1,91 @@
...
     if [ "${CI_RUNNING}" == "true" ]; then
         if [ -z "${WIFI_INTERFACE}" ]; then
             WIFI_INTERFACE="CI TEST INTERFACE"
         fi
-        if [ -z "${WIFI_REGION}" ]; then
-            WIFI_REGION="CI TEST REGION"
-        fi
+        # if [ -z "${WIFI_REGION}" ]; then
+        #     WIFI_REGION="CI TEST REGION"
+        # fi
```
[the remark this answers is not in the document]
Comment (AlvinSchiller) [reply installation/routines/setup_autohotspot.sh:39]: Still a draft, will be removed

File: installation/routines/setup_autohotspot_NetworkManager.sh [hunk 8]
```diff
@@ -0,0 +1,101 @@
+
```
[the remark this answers is not in the document]
Comment (AlvinSchiller) [reply installation/routines/setup_autohotspot_NetworkManager.sh:1]: will add it

File: installation/routines/setup_autohotspot.sh (hunk 6 above)
Comment (AlvinSchiller) [reply installation/routines/setup_autohotspot.sh:80]: Its for compatibility.
Bullseye supports NetworkManager, but its not default.
The user can change it (similar to the description in https://github.com/MiczFlor/RPi-Jukebox-RFID/issues/2184), but its a manual task.

It also not trivial to automate it, as the network will reset and the connection will be lost.
Also there would be a mapper from WPA entries needed to activate wifi again, if the user has only wifi connection (which will be in most cases i suppose).

File: installation/routines/setup_autohotspot.sh [hunk 9]
```diff
@@ -1,119 +1,91 @@
...
 
-# inspired by
-# https://www.raspberryconnect.com/projects/65-raspberrypi-hotspot-accesspoints/158-raspberry-pi-auto-wifi-hotspot-switch-direct-connection
+_is_dhcpcd_enabled() {
+	echo $(_is_service_enabled "dhcpcd.service")
+}
 
+_is_NetworkManager_enabled() {
+	echo $(_is_service_enabled "NetworkManager.service")
+}
```
[the remark this answers is not in the document]
Comment (AlvinSchiller) [reply installation/routines/setup_autohotspot.sh:24]: ?

File: installation/routines/customize_options.sh [hunk 10]
```diff
@@ -49,51 +49,101 @@ Do you want to disable IPv6? [Y/n]"
...
+                        read -r response_pw
+                        response_pw_length=$(get_string_length ${response_pw})
+                    done
+
+                    local response_ip=""
+                    while [[ ! "$response_ip" =~ ^((25[0-5]|(2[0-4]|1[0-9]|[1-9]|)[0-9])\.){3}((25[0-5]|(2[0-4]|1[0-9]|[1-9]|)[0-9]))$ ]]
+                    do
+                        print_c "Please type the hotspot static IP (e.g. 10.0.0.5, 192.168.1.199):"
+                        read -r response_ip
+                    done
```
[the remark this answers is not in the document]
Comment (AlvinSchiller) [reply installation/routines/customize_options.sh:107]: we provide a default the user can stick to, but why not give the chance to make changes?
This is the feature set v2 has already.

File: installation/includes/01_default_config.sh [hunk 11]
```diff
@@ -4,8 +4,10 @@ BUILD_LIBZMQ_WITH_DRAFTS_ON_DEVICE=${BUILD_LIBZMQ_WITH_DRAFTS_ON_DEVICE:-"false"
 ENABLE_STATIC_IP=true
 DISABLE_IPv6=true
 ENABLE_AUTOHOTSPOT=false
-AUTOHOTSPOT_CHANGE_PASSWORD=false
+AUTOHOTSPOT_SSID="Phoniebox_Hotspot"
 AUTOHOTSPOT_PASSWORD="PlayItLoud!"
+AUTOHOTSPOT_IP="10.0.0.5"
```
[the remark this answers is not in the document]
Comment (AlvinSchiller) [reply installation/includes/01_default_config.sh:9]: Its the hard coded IP set in the tutorial / scripts used as base.
It just wasn't changed.

File: installation/routines/customize_options.sh [hunk 12]
```diff
@@ -49,51 +49,101 @@ Do you want to disable IPv6? [Y/n]"
...
-      esac
+            ;;
+    esac
+
+    if [ "$ENABLE_AUTOHOTSPOT" = true ]; then
+        local response_autohotspot
+        while [[ $response_autohotspot != "n" ]]
+        do
+            print_c "
+--- Current configuration for Autohotpot
```
Comment (Groovylein) [inline installation/routines/customize_options.sh:75]: I like this approach! 👍

File: installation/routines/customize_options.sh [hunk 13]
```diff
@@ -49,51 +49,101 @@ Do you want to disable IPv6? [Y/n]"
...
+                    local response_pw_length=0
+                    while [[ $response_pw_length -lt 8 || $response_pw_length -gt 63 ]]
+                    do
+                        print_c "Please type the new password (must be between 8 and 63 characters long):"
+                        read -r response_pw
+                        response_pw_length=$(get_string_length ${response_pw})
+                    done
+
+                    local response_ip=""
+                    while [[ ! "$response_ip" =~ ^((25[0-5]|(2[0-4]|1[0-9]|[1-9]|)[0-9])\.){3}((25[0-5]|(2[0-4]|1[0-9]|[1-9]|)[0-9]))$ ]]
```
Comment (Groovylein) [inline installation/routines/customize_options.sh:103]: Should we really let people decide about the IP? Most of them don't have the knowledge I assume.

I would keep it as default.

File: installation/routines/customize_options.sh [hunk 14]
```diff
@@ -49,51 +49,101 @@ Do you want to disable IPv6? [Y/n]"
...
-      fi
-  fi
 
-  log "ENABLE_AUTOHOTSPOT=${ENABLE_AUTOHOTSPOT}"
-  if [ "$ENABLE_AUTOHOTSPOT" = true ]; then
-    log "AUTOHOTSPOT_PASSWORD=${AUTOHOTSPOT_PASSWORD}"
-  fi
+        if [ "$ENABLE_STATIC_IP" = true ]; then
+            print_c "
+Static IP cannot be enabled with WiFi hotspot.
```
Comment (Groovylein) [inline installation/routines/customize_options.sh:133]: I would give this information when asking if Autohotspot  should be enabled.

File: installation/routines/customize_options.sh (hunk 10 above)
[the remark this answers is not in the document]
Comment (pabera) [reply installation/routines/customize_options.sh:107]: In my opinion, there is no obvious reason, as there are no competing IP addresses within the hotspot. 

Unless you don't offer a tool to change it again, I would just define a default. 

Imagine, what happens if you have forgotten the IP address you have set manually? And you don't have access to SSH at this moment. You can't even look up the Phoniebox docs as they won't help.

That can already be a problem with the password, but here I agree it must be changeable.

File: installation/includes/01_default_config.sh [hunk 15]
```diff
@@ -4,8 +4,10 @@ BUILD_LIBZMQ_WITH_DRAFTS_ON_DEVICE=${BUILD_LIBZMQ_WITH_DRAFTS_ON_DEVICE:-"false"
 ENABLE_STATIC_IP=true
 DISABLE_IPv6=true
 ENABLE_AUTOHOTSPOT=false
-AUTOHOTSPOT_CHANGE_PASSWORD=false
+AUTOHOTSPOT_SSID="Phoniebox_Hotspot"
```
[the remark this answers is not in the document]
Comment (AlvinSchiller) [reply installation/includes/01_default_config.sh:7]: Another idea would be, to use the hostname as addition for the ssid (thats also the default of the NetworkManager if no ssid is provided).
Is it realistic enough that people change the hostname if they use multiple Phonieboxes?

**Comment by AlvinSchiller** [comment]
IPv6 disablement was also broken for bullseye. 
Changed it to a kernel parameter to thoroughly disable ipv6 also for future connections.

File: installation/includes/02_helpers.sh [hunk 16]
```diff
@@ -119,6 +128,34 @@ download_from_url() {
     return $?
 }
 
+get_string_length() {
+    local string="$1"
+    # "-n" option is needed otherwise an additional linebreak char is added by echo
+    echo -n ${string} | wc -m
+}
+
+is_service_enabled() {
```
Comment (pabera) [inline installation/includes/02_helpers.sh:137]: There is quite some similarity with `verify_service_enablement()`. Could those 2 functions be combined?

File: installation/routines/set_raspi_config.sh [hunk 17]
```diff
@@ -17,13 +17,15 @@ _run_set_raspi_config() {
...
-      sudo cp "${RPI_BOOT_CONFIG_FILE}" "${DISABLE_ONBOARD_AUDIO_BACKUP}"
-      sudo sed -i "s/^\(dtparam=\([^,]*,\)*\)audio=\(on\|true\|yes\|1\)\(.*\)/\1audio=off\4/g" "${RPI_BOOT_CONFIG_FILE}"
+    if grep -q -E "^dtparam=([^,]*,)*audio=(on|true|yes|1).*" "${configFile}" ; then
+      local configFile_backup="${configFile}.backup.audio_on_$(date +%d.%m.%y_%H.%M.%S)"
+      log "    Backup ${configFile} --> ${configFile_backup}"
+      sudo cp "${configFile}" "${configFile_backup}"
+      sudo sed -i "s/^\(dtparam=\([^,]*,\)*\)audio=\(on\|true\|yes\|1\)\(.*\)/\1audio=off\4/g" "${configFile}"
     else
-      log "    On board audio seems to be off already. Not touching ${RPI_BOOT_CONFIG_FILE}"
+      log "    On board audio seems to be off already. Not touching ${configFile}"
```
Comment (pabera) [inline installation/routines/set_raspi_config.sh:28]: This is not related to NetworkManager. Since I already worked on this in my [PR](https://github.com/pabera/RPi-Jukebox-RFID/pull/2/files#diff-8ef83323c04590c1500b81c1a09100d8b892d9833f77587112fa20e44d34beab), maybe you could remove it to avoid conflicts

File: installation/routines/setup_autohotspot.sh [hunk 18]
```diff
@@ -1,119 +1,70 @@
...
-}
-
-_install_service_and_timer() {
-    sudo cp "${INSTALLATION_PATH}"/resources/autohotspot/autohotspot.service /etc/systemd/system/autohotspot.service
-    sudo systemctl enable autohotspot.service
-
-    local cron_autohotspot_file="/etc/cron.d/autohotspot"
-    sudo cp "${INSTALLATION_PATH}"/resources/autohotspot/autohotspot.timer "${cron_autohotspot_file}"
-    sudo sed -i "s|%%USER%%|${CURRENT_USER}|g" "${cron_autohotspot_file}"
+_config_file_backup() {
```
Comment (pabera) [inline installation/routines/setup_autohotspot.sh:22]: Shouldn't that be a helper function?

File: installation/routines/setup_autohotspot.sh [hunk 19]
```diff
@@ -1,119 +1,70 @@
...
+        sudo touch "${config_flag_file}"
+    elif [ ! -f "${config_orig_file}" ] && [ ! -f "${config_flag_file}" ]; then
+        sudo cp "${config_file}" "${config_orig_file}"
+    fi
 }
 
-_install_autohotspot_script() {
-    sudo cp "${INSTALLATION_PATH}"/resources/autohotspot/autohotspot "${AUTOHOTSPOT_TARGET_PATH}"
-    sudo chmod +x "${AUTOHOTSPOT_TARGET_PATH}"
+_config_file_revert() {
```
Comment (pabera) [inline installation/routines/setup_autohotspot.sh:35]: Same here?

**Review by pabera — COMMENTED** [review]
Couldn't install it, but overall looks good! Will approve once tested on a Pi

File: installation/routines/setup_autohotspot.sh (hunk 18 above)
Comment (AlvinSchiller) [reply installation/routines/setup_autohotspot.sh:22]: was pretty special to this setup. I dont know if it fits also other cases.
But can pull it down and add some docs

File: installation/routines/set_raspi_config.sh (hunk 17 above)
Comment (AlvinSchiller) [reply installation/routines/set_raspi_config.sh:28]: Can do, but this leaves this feature not functional for bookworm and some clutter.

It should also not be a problem with your branch, as you deleted the whole file, which can just be kept on merging.
I can support on merging if you like.

File: installation/includes/02_helpers.sh (hunk 16 above)
Comment (AlvinSchiller) [reply installation/includes/02_helpers.sh:150]: Done. Extracted the repetitive code to an internal function

File: documentation/builders/installation.md [hunk 20]
```diff
@@ -81,11 +80,11 @@ You will need a terminal, like PuTTY for Windows or the Terminal app for Mac to
...
+#### Network management since Bookworm
 <details>
-With Bookworm the network settings have changed. Now "NetworkManager" is used instead of "dhcpcd".
-This breaks breaks network related features like "Static IP", "Wifi Setup" and "Autohotspot".
-Before running the installation, the network config has to be changed via raspi-config, to use the "old" dhcpcd network settings.
+With Bookworm the network management has changed. Now "NetworkManager" is used instead of "dhcpcd".
+Both ways are supported in the installation, but "NetworkManager" is recommended as it's simpler to setup and use. 
+For Bullseye this can also be activated, though it's a manual process before running the installation.
 
 :warning:
```
[the remark this answers is not in the document]
Comment (AlvinSchiller) [reply documentation/builders/installation.md:89]: Alerts are not rendered inside of the `details`

See https://github.com/MiczFlor/RPi-Jukebox-RFID/pull/2216#discussion_r1454236302

**Review by s-martin — APPROVED** [review]
